neqo-transport/src/frame.rs

```diff
@@ -263,43 +263,85 @@ impl Frame {
         fin: bool,
         space: usize,
     ) -> Option<(Frame, usize)> {
-        let mut remaining = space.saturating_sub(1 + Encoder::varint_len(stream_id));
+        let mut overhead = 1 + Encoder::varint_len(stream_id);
         if offset > 0 {
-            remaining = remaining.saturating_sub(Encoder::varint_len(offset));
+            overhead += Encoder::varint_len(offset);
+        }
+
+        if overhead > space {
+            qdebug!(
+                "Frame::new_stream -> None; hdrlen {} > space {}",
+                overhead,
+                space
+            );
+            return None;
         }
-        let (fin, fill) = if data.len() > remaining {
-            if remaining == 0 {
-                return None;
-            }
 
+        let (fin, fill) = if data.len() > (space - overhead) {
             // More data than fits, fill the packet and negate |fin|.
             (false, true)
-        } else if data.len() == remaining {
+        } else if data.len() == (space - overhead) {
+            if data.is_empty() && !fin {
+                // Do not frame unless space for some data or includes FIN
+                qdebug!(
+                    "Frame::new_stream -> None; hdrlen {} == space {} and !fin",
+                    overhead,
+                    space
+                );
+            }
             // Exact fit, fill the packet, keep |fin|.
             (fin, true)
         } else {
             // Too small, so include a length.
-            let data_len = min(remaining - 1, data.len());
-            remaining -= Encoder::varint_len(u64::try_from(data_len).unwrap());
-            remaining = min(data.len(), remaining);
-            // In case the added length causes this to spill over, check |fin| again.
-            (fin && remaining == data.len(), false)
+            let data_len = min(space - overhead - 1, data.len());
+            overhead += Encoder::varint_len(u64::try_from(data_len).unwrap());
+            if overhead > space {
+                // Not even enough space for length field
+                qdebug!(
+                    "Frame::new_stream -> None; hdrlen w/len {} > space {}",
+                    overhead,
+                    space
+                );
+                return None;
+            }
+            if overhead == space && !fin {
```

Comment:
I think that you could consolidate this with the check on the exact fit branch by moving this below.  You can also consolidate the `overhead > space` case as well:

```rust
if overhead > space || (overhead == space && !fin) {
  return None;
}
```

Then you only have one check.  The cost is that you execute a little more code in those cases where this code is hit.  You might also consolidate further:


```rust
let empty_frame = overhead == space || data.is_empty();
if overhead > space || (empty_frame && !fin) {
  return None;
}
```